### ISSUE TITLE
Some dynamic formats, canonical hash not matching pot entry

### ISSUE THREAD

**Issue by magnumripper** [openwall/john#3124]
Here's dynamic_1034: md5($p.$u) username "jumbo" password "magnum"
```
$ echo 'jumbo:$dynamic_1034$16ccb184ee1f16dd9af509135f504442' > test.in
```
Here's what the format reports as the canonical hash (the second, `$$Ujumbo` one):
```
$ ../run/john --show=types test.in
jumbo:$dynamic_1034$16ccb184ee1f16dd9af509135f504442:::/:/:/:dynamic_1034:0:1:0:$dynamic_1034$16ccb184ee1f16dd9af509135f504442$$Ujumbo:0:
```

As expected, the same canonical hash is given for a bare hash and specifying format on command line instead:
```
$ echo 'jumbo:16ccb184ee1f16dd9af509135f504442' > test.in

$ ../run/john --show=types test.in -form:dynamic_1034
jumbo:16ccb184ee1f16dd9af509135f504442:::/:/:/:dynamic_1034:0:1:0:$dynamic_1034$16ccb184ee1f16dd9af509135f504442$$Ujumbo:0:
```

Fine so far. But here's what's actually in the pot file after cracking:
```
$ ../run/john test.in -mask:magnum
Using default input encoding: UTF-8
Loaded 1 password hash (dynamic_1034 [md5($p.$u) (PostgreSQL MD5) 256/256 AVX2 8x3])
Press 'q' or Ctrl-C to abort, almost any other key for status
magnum           (jumbo)
1g 0:00:00:00  10.00g/s 10.00p/s 10.00c/s 10.00C/s magnum
Use the "--show" option to display all of the cracked passwords reliably
Session completed

$ cat ../run/john.pot
$dynamic_1034$16ccb184ee1f16dd9af509135f504442$HEX$24556a756d626f:magnum
```
They don't match, the canonical format isn't canonical. How come? How can this even happen!? External tools should rely on the `--show=types` information and to the best of my knowledge they can, except for dynamic formats.

And it gets worse: Let's say I *did* hexify the input file, in order to make some external tool recognize the pot entries as matching the input file:
```
$ cat test.in
jumbo:$dynamic_1034$16ccb184ee1f16dd9af509135f504442$HEX$24556a756d626f

$ ../run/john --show=types test.in
jumbo:$dynamic_1034$16ccb184ee1f16dd9af509135f504442$HEX$24556a756d626f:::/:/:/:dynamic_1034:0:1:0:$dynamic_1034$16ccb184ee1f16dd9af509135f504442$$Ujumbo:0:
```
In this case, dynamic *reverses* the hexifying. So no matter how I do it, the canonical hash from a `--show=types` of the input file doesn't match what is written in the pot file.

I assume the problem is you are doing the hexify stuff somewhere outside of prepare().

**Comment by jfoug** [comment]
This has much in common with #2437  The whole issue needs to be stepped back from, and fully dug into, so that fixing one item, does not break others

**Comment by magnumripper** [comment]
Fixed by #4947
